samples/azure_attestation/ra_client/client.go

```diff
@@ -42,49 +47,30 @@ func main() {
 	// Get cerificate for token verification from attestation provider.
 	tlsConfig = &tls.Config{}
 	jwkSetBytes := httpGet(tlsConfig, issAddr+"/certs")
-	var keySet JSONWebKeySet
-	if err := json.Unmarshal(jwkSetBytes, &keySet); err != nil {
-		panic(err)
-	}
+	keySet := mustParseKeySet(jwkSetBytes)
 	fmt.Printf("🆗 Loaded attestation provider's verification keys from %s.\n", issAddr+"/certs")
 
 	// Parse token.
-	parts := strings.Split(string(tokenBytes), ".")
-	for i, part := range parts {
-		p, err := base64.RawURLEncoding.DecodeString(part)
-		if err != nil {
-			panic(err)
-		}
-		parts[i] = string(p)
-	}
-	var header JOSEHeader
-	if err := json.Unmarshal([]byte(parts[0]), &header); err != nil {
+	token, err := jwt.ParseSigned(string(tokenBytes))
+	if err != nil {
 		panic(err)
 	}
 
-	// Find signing key.
-	key, err := keySet.findKid(header.Kid)
-	if err != nil {
+	// Verify token and get claims.
+	var publicClaims jwt.Claims
+	var privateClaims privateClaims
+	if err := token.Claims(&keySet, &publicClaims, &privateClaims); err != nil {
 		panic(err)
 	}
-	fmt.Println("🆗 Found the attestation provider's key which was used to sign the token.")
-
-	// Convert key to PEM, since the provided key of this instance are not valid JSON Web Keys
-	pubk := jwkToRSAPub(key)
+	fmt.Println("✅ Token is valid.")
```

Comment:
I think this is misplaced here, since token.Claims does not provide any new validation. The tokens signature was already validated with the jwt.ParseSigned() on line 54. Previous the claims.Valid() function checked the validity of the time based information inside the token. Since the enclave held data we're using is a cert, replay attacks should not be a problem. However, this is a special case and in any other situation, at least the validation of replay protecting claims should be performed.

I would suggest moving this next to the jwt.ParseSigned() call and clearify like "Token signature is valid".

Reply:
The functions are poorly named. I think ParseSigned() does no validation (it does not take a key) and Claims() validates because it takes a key and there also is a function UnsafeClaimsWithoutVerification(). The time based information is validated on line 94 with publicClaims.Validate.

Reply:
Okay, I think you're right regarding the signature validation.
But at line 94, you only validate the issuer, as far as I understand the docs. The [documentation of `Expected`](https://pkg.go.dev/gopkg.in/square/go-jose.v2@v2.5.1/jwt#Expected) states: "Expected defines values used for protected claims validation. If field has zero value then validation is skipped.", and you don't set the `Time` value.

Reply:
Good catch! Can you fix this?